### PR TITLE
feat(threaded-replies): create base of ActivityThread component

### PR DESCRIPTION
**Pull request by wpiesiak**
create base of ActivityThread component to support threads for Comments and Annotation. 

The role of this component is to group an activity (Comment or Annotation) with it’s replies

Props:
- replies?: Comment[]
- repliesTotal: number
- onReplyCreate?: Function
- onGetReplies?: Function

UI should be wrapped with ActivityCard component so that the entire thread is treated as a card
UI will render the main item based on it’s type using Comment or AnnotationActivity

### CODE REVIEW

File: src/common/types/annotations.js [hunk 1]
```diff
@@ -47,7 +47,8 @@ export type Annotation = {
     modified_at: string,
     modified_by: User,
     permissions: AnnotationPermission,
-    replies?: Array<Reply>,
+    replies?: Array<Comment>,
+    repliesTotal?: number,
```
Comment (jstoffan) [inline src/common/types/annotations.js:51]: Nearly all V2 API fields use snake_case naming (see the other fields on this object), so this would be `replies_total`.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js [hunk 2]
```diff
@@ -85,6 +88,35 @@ const ActiveState = ({
...
+                        >
+                            <ActivityThread
+                                currentFileVersionId={currentFileVersionId}
+                                getAvatarUrl={getAvatarUrl}
+                                isCurrentVersion={isCurrentVersion}
+                                item={item}
+                                onAnnotationDelete={onAnnotationDelete}
+                                onAnnotationEdit={onAnnotationEdit}
+                                onAnnotationSelect={onAnnotationSelect}
+                                // onAnnotationReplyCreate={() => {}} // TODO will be implemented in next PR
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:110]: We generally want to avoid commented code, even if it's temporary. Can we omit these lines?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js [hunk 3]
```diff
@@ -85,6 +88,35 @@ const ActiveState = ({
...
+                            className="bcs-activity-thread"
+                            data-testid="activity-thread"
+                            isFocused={isFocused}
+                            ref={refValue}
+                        >
+                            <ActivityThread
+                                currentFileVersionId={currentFileVersionId}
+                                getAvatarUrl={getAvatarUrl}
+                                isCurrentVersion={isCurrentVersion}
+                                item={item}
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:106]: It looks like the other activity types spread `{...item}` onto their props. Should we do the same here for consistency?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+const ActivityThread = ({
+    currentUser,
+    getAvatarUrl,
+    getMentionWithQuery,
+    getUserProfileUrl,
+    isCurrentVersion,
+    item,
+    mentionSelectorContacts,
+    // eslint-disable-next-line no-unused-vars
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js:47]: Can we omit these props for now?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    onGetAnnotationReplies = noop,
+    // eslint-disable-next-line no-unused-vars
+    onGetCommentReplies = noop,
+    translations,
+}: Props) => {
+    return (
+        <ActivityCard className="bcs-ActivityThread">
+            {item.type === 'comment' && (
+                <Comment
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js:65]: It doesn't look like `Comment` and `AnnotationActivity` share much behavior. Can we move these conditions up to `ActiveState` and allow `ActivityThread` to accept them as a `children` prop?

File: src/elements/content-sidebar/activity-feed/activity-feed/__tests__/ActivityThread.test.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import { shallow, ShallowWrapper } from 'enzyme';
+import Comment from '../../comment/Comment';
+import AnnotationActivity from '../../annotations/AnnotationActivity';
+import ActivityThread from '../ActivityThread.js';
+import ActivityCard from '../../ActivityCard';
+
+describe('src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread', () => {
+    // eslint-disable-next-line flowtype/no-types-missing-file-annotation
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/__tests__/ActivityThread.test.js:9]: We should try to avoid disabling linting rules whenever possible. Should this file be marked with a Flow annotation

File: src/common/types/feed.js [hunk 7]
```diff
@@ -40,6 +40,8 @@ type Comment = {
     message?: string,
     modified_at: string,
     permissions: BoxCommentPermission,
+    replies?: Array<Comment>,
+    repliesTotal?: number,
```
Comment (ConradJChan) [inline src/common/types/feed.js:44]: Will the new reply be exactly the `Comment` type as it is today or will the type be slightly different? Also, we could consider creating a common type for this to be used in `Comment` and `Annotation`

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js (hunk 5 above)
Comment (ConradJChan) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js:65]: +1, it may be easier to use and extend in the future (for Tasks maybe) if we make this `ActivityThread` component more composable, something like 
```
<ActivityThread>
    <Comment />
</ActivityThread>
```
where it's only responsibility is really to show the reply thread in it's compact form or expanded form

File: src/common/types/feed.js [hunk 8]
```diff
@@ -39,8 +39,14 @@ type Comment = {
     is_reply_comment?: boolean,
     message?: string,
     modified_at: string,
+    parent?: {
+        id: string,
+        type: 'commnet' | 'annotation',
```
Comment (jstoffan) [inline src/common/types/feed.js:44]: Typo.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js [hunk 9]
```diff
@@ -161,6 +193,32 @@ const ActiveState = ({
...
+                        if (hasReplies) {
+                            return (
+                                <ActivityItem
+                                    key={item.type + item.id}
+                                    className="bcs-activity-thread"
+                                    data-testid="activity-thread"
+                                    isFocused={isFocused}
+                                    ref={refValue}
+                                >
+                                    <ActivityThread {...item}>
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:205]: Do we need to spread `{...item}` here or can it be omitted?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
+// @flow
+import * as React from 'react';
+import ActivityCard from '../ActivityCard';
+
+import type { Comment } from '../../../../common/types/feed';
+
+type Props = {
+    children: React.Node,
+    id: string,
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js:9]: It looks like most of these props are unused right now. Can they be removed?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js (hunk 9 above)
Comment (wpiesiak) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js]: It can be omitted currently and added in next PR 👍

File: src/common/types/feed.js (hunk 7 above)
Comment (wpiesiak) [reply src/common/types/feed.js]: Every reply will be basically a comment from api perspective

**Review by jstoffan — APPROVED** [review]

File: src/common/types/feed.js [hunk 11]
```diff
@@ -39,8 +39,14 @@ type Comment = {
     is_reply_comment?: boolean,
     message?: string,
     modified_at: string,
+    parent?: {
+        id: string,
+        type: 'comment' | 'annotation',
+    },
     permissions: BoxCommentPermission,
+    replies?: Array<Comment>,
     tagged_message: string,
```
Comment (jstoffan) [inline src/common/types/feed.js:48]: Will these unused fields be included in the new schema? Or should we be using a new "Comment" type for these flows?

Comment (jstoffan) [reply src/common/types/feed.js:48]: I don't think these fields are included in our new backend schema for these objects, though. We should have accurate types, even if it's a bit more work.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js [hunk 12]
```diff
@@ -88,6 +91,35 @@ const ActiveState = ({
 
                 switch (item.type) {
                     case 'comment':
+                        if (hasReplies) {
```
Comment (ConradJChan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:94]: I'm wondering if it would be cleaner if we deferred the `hasReplies` check in the `ActivityThread` component

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js [hunk 13]
```diff
@@ -91,28 +94,31 @@ const ActiveState = ({
                         return (
                             <ActivityItem
                                 key={item.type + item.id}
-                                className="bcs-activity-feed-comment"
-                                data-testid="comment"
+                                className={hasReplies ? 'bcs-activity-thread' : 'bcs-activity-feed-comment'}
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:97]: It doesn't look like `.bcs-activity-thread` is used at the moment. Will we need different styles for the ActivityItem component in the future?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js [hunk 14]
```diff
@@ -91,28 +94,31 @@ const ActiveState = ({
...
+                                data-testid={hasReplies ? 'activity-thread' : 'comment'}
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:98]: Should this be on the `ActivityThread` component, instead?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+import ActivityCard from '../ActivityCard';
+
+type Props = {
+    children: React.Node,
+    hasReplies: boolean,
+};
+
+const ActivityThread = ({ children, hasReplies }: Props) => {
+    if (hasReplies) {
+        return <ActivityCard className="bcs-ActivityThread">{children}</ActivityCard>;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js:12]: It looks like this will create a nested ActivityCard output, since the `Comment` component's root is also ActivityCard. Is that desirable? It also doesn't look like `.bcs-ActivityThread` is used at the moment. Will it be in the future?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js (hunk 13 above)
Comment (wpiesiak) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:97]: Probably we won't need to style this ActivityItem differently. I was just following the pattern in the component, but noticed  'bcs-activity-feed-comment' is not used neither

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js (hunk 15 above)
Comment (wpiesiak) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js:12]: I am aware of it, but I think it would be best to treat ActivityThread as ActivityCard containing main comment with replies and wrapping it up. I think it shouldn't be a problem to have nested ActivityCard for parent comment and each reply. Probably it would just need style fixes for the spacing for the parent component but IMO it would look like more consistent.

Comment (wpiesiak) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js:12]: And I would keep the className - it will be used in next [PR](https://github.com/box/box-ui-elements/pull/3010/files)

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js (hunk 14 above)
Comment (jstoffan) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:97]: I ask because ActivityItem already has a `data-testid` that we can use. If we want to select on ActivityThread, as well, we can add a `data-testid` there.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js [hunk 16]
```diff
@@ -164,25 +169,28 @@ const ActiveState = ({
                         return (
                             <ActivityItem
                                 key={item.type + item.id}
-                                className="bcs-activity-feed-annotation-activity"
-                                data-testid="annotation-activity"
+                                className={!hasReplies && 'bcs-activity-feed-annotation-activity'}
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:172]: Can we remove this logic? The class seems unrelated since it's on a parent of the ActivityThread.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js [hunk 17]
```diff
@@ -164,25 +169,28 @@ const ActiveState = ({
...
+                                data-testid={hasReplies ? 'activity-thread' : 'annotation-activity'}
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js:173]: Again, this should be on the `ActivityThread`, as the name implies.

File: src/common/types/feed.js (hunk 11 above)
Comment (jstoffan) [reply src/common/types/feed.js:48]: Did we come to a resolution on this question?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js (hunk 14 above)
Comment (wpiesiak) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActiveState.js]: No I get it 😄 Sounds good - fixed

**Review by jstoffan — APPROVED** [review]
